dust-lambda-extension/extensions/dust

```diff
@@ -0,0 +1,294 @@
+
+	rm -f "$temp_file"
+
+	if [[ -z $EXTENSION_ID ]]; then
+		LOG_ERROR "Failed to register extension or parse Extension ID."
+		exit 1
+	fi
+}
+
+register_extension() {
+	# Only register for SHUTDOWN events (INVOKE events are not needed)
+	local registration_payload='{"events":["SHUTDOWN"]}'
+
+	# Make a single registration request with headers included
+	local temp_file
+	temp_file=$(mktemp)
+	local http_code
+
+	http_code=$(curl --silent --write-out "%{http_code}" \
+		--dump-header "$temp_file" \
+		--request POST "$REGISTER_API" \
+		--header "Lambda-Extension-Name: $EXTENSION_NAME" \
+		--header "Content-Type: application/json" \
+		--data "$registration_payload" \
+		--output /dev/null)
+
+	if [[ $http_code -ne 200 ]]; then
+		LOG_ERROR "Failed to register extension. HTTP status: $http_code"
+		LOG_DEBUG "Response headers: $(cat "$temp_file")"
+		rm -f "$temp_file"
+		exit 1
+	fi
+
+	extract_extension_id "$temp_file"
+
+	LOG_INFO "Extension registered with ID: $EXTENSION_ID"
+}
+
+# Main execution starts here
+main() {
+	LOG_INFO "Starting Lambda extension: $EXTENSION_NAME"
+
+	# Set up signal handlers
+	trap handle_signal SIGINT SIGTERM
+
+	register_extension
+
+	# Setup chalk binary
+	if ! setup_chalk_binary; then
+		LOG_ERROR "Failed to setup chalk binary"
+		exit 1
+	fi
+
+	# Process chalk marks
+	process_chalk_marks || LOG_WARNING "Chalk mark processing failed, continuing anyway"
+
+	# Poll for events from the Lambda Runtime API
+	LOG_INFO "Starting event polling loop."
+
+	local consecutive_failures=0
+	local max_failures=$MAX_FAILURES
+
+	while [[ $EXIT_REQUESTED == "false" ]]; do
+		LOG_DEBUG "Polling for next event..."
+
+		local event_response
+		if event_response=$(make_api_request "GET" "$NEXT_API" "Lambda-Extension-Identifier: $EXTENSION_ID"); then
+			consecutive_failures=0
+
+			# Extract event type from JSON response
+			local event_type
+			event_type=$(extract_json_field "$event_response" "eventType")
+
+			LOG_DEBUG "Received event: $event_type"
+
+			case "$event_type" in
+			"SHUTDOWN")
+				LOG_INFO "Received SHUTDOWN event. Initiating clean shutdown."
+				EXIT_REQUESTED=true
+				do_shutdown
+				;;
+			"INVOKE")
+				LOG_DEBUG "Received INVOKE event. Continuing to poll."
+				;;
+			"")
+				LOG_WARNING "Received event with no eventType field"
+				LOG_DEBUG "Full event response: $event_response"
+				;;
+			*)
+				LOG_DEBUG "Received unknown event type: $event_type. Continuing to poll."
+				;;
+			esac
+		else
+			# Increment failure counter
+			((consecutive_failures++))
+			LOG_ERROR "Failed to get next event from Runtime API (failure $consecutive_failures/$max_failures)"
+
+			if [[ $consecutive_failures -ge $max_failures ]]; then
+				LOG_ERROR "Too many consecutive failures. Shutting down."
+				EXIT_REQUESTED=true
+				do_shutdown
+			fi
+
+			# Brief backoff before retrying
+			sleep 1
+		fi
+	done
+}
+
+# Run main function if script is executed directly
+if [[ ${BASH_SOURCE[0]} == "${0}" ]]; then
```

Comment:
do we expect people to source this script?

checking `BASH_SOURCE` is needed to separate `source foo.sh` vs `./foo.sh` but if its not a concern here usually shell scripts just execute directly whatever they need